src/libraries/System.Private.CoreLib/src/System/Utf8String.Construction.cs

```diff
@@ -256,6 +264,96 @@ internal static Utf8String CreateFromRune(Rune value)
+                    return null;
+                }
+
+                // At this point we know transcoding succeeded, so the original input data was well-formed.
+                // We'll memcpy the scratch buffer into the new Utf8String instance, which is very fast.
+
+                newBuffer = new byte[scratchBytesWritten + 1]; // null-terminated
+                scratch.Slice(0, scratchBytesWritten).CopyTo(newBuffer);
+                return newBuffer;
+            }
+
+            // First, determine how many UTF-8 bytes we'll need in order to represent this data.
+            // This also checks the input data for well-formedness.
+
+            long utf8CodeUnitCountAdjustment;
+
+            unsafe
+            {
+                fixed (char* pChars = &MemoryMarshal.GetReference(value))
+                {
+                    if (Utf16Utility.GetPointerToFirstInvalidChar(pChars, value.Length, out utf8CodeUnitCountAdjustment, out int _) != (pChars + (uint)value.Length))
+                    {
+                        return null;
+                    }
+                }
+            }
+
+            // The max possible expansion transcoding UTF-16 to UTF-8 is that each input char corresponds
+            // to 3 UTF-8 bytes. This is most common in CJK languages. Since the input buffer could be
+            // up to int.MaxValue elements in length, we need to use a 64-bit value to hold the total
+            // required UTF-8 byte length. However, the VM places restrictions on how large a Utf8String
+            // instance can be, and the maximum allowed element count is just under int.MaxValue. (This
+            // mirrors the restrictions already in place for System.String.) The VM will throw an
+            // OutOfMemoryException if anybody tries to create a Utf8String instance larger than that,
+            // so if we detect any sort of overflow we'll end up passing int.MaxValue down to the allocation
+            // routine. This normalizes the OutOfMemoryException the caller sees.
+
+            long totalUtf8BytesRequired = (uint)value.Length + utf8CodeUnitCountAdjustment;
+            if (totalUtf8BytesRequired > int.MaxValue)
+            {
+                totalUtf8BytesRequired = int.MaxValue;
+            }
+
+            // We can get away with FastAllocateSkipZeroInit here because we're not going to return the
+            // new Utf8String instance to the caller if we don't overwrite every byte of the buffer.
+
+            newBuffer = new byte[(int)totalUtf8BytesRequired + 1]; // null-terminated
+
+            // Now transcode the UTF-16 input into the newly allocated Utf8String's buffer. We can't call the
+            // "skip validation" transcoder because the caller could've mutated the input buffer between the
+            // initial counting step and the transcoding step below.
+
+            status = Utf8.FromUtf16(value, newBuffer, out _, out int bytesWritten, replaceInvalidSequences: false);
```

Comment:
Pass `newBuffer[..^1]` instead of `newBuffer`. Otherwise the input could overwrite the terminating null character, which would cause problems when somebody attempts to pin this instance.